components/fxa-client/src/oauth.rs

```diff
@@ -172,25 +172,47 @@ impl FirefoxAccount {
         &self,
         auth_params: AuthorizationParameters,
     ) -> Result<String> {
-        // TODO: Add validation to check for which keys the client is allowed to request,
-        //       And if we have those keys stored locally.
-        //       Currently we pull ALL available keys that match the requested scopes, regardless
-        //       if the client is actually allowed to request them.
         let session_token = self.get_session_token()?;
+
+        // Validate request to ensure that the client is actually allowed to request
+        // the scopes they requested
+        let valid_scopes = self.client.scoped_key_data(
+            &self.state.config,
+            &session_token,
+            &auth_params.client_id,
+            &auth_params.scope,
+        )?;
+
+        if let Some(not_allowed_scope) = auth_params
+            .scope
+            .split_whitespace()
+            .find(|scope| !valid_scopes.contains_key(*scope))
+        {
+            return Err(ErrorKind::NotAllowedToRequestScope(not_allowed_scope.to_string()).into());
+        }
+
         let keys_jwe = if let Some(keys_jwk) = auth_params.keys_jwk {
             let keys_jwk = base64::decode_config(keys_jwk, base64::URL_SAFE_NO_PAD)?;
             let keys_jwk = String::from_utf8(keys_jwk)?;
-            let requested_scopes: HashSet<&str> = auth_params.scope.split_whitespace().collect();
-            let data: HashMap<String, ScopedKey> = self
+            let scoped_keys: HashMap<String, ScopedKey> = self
                 .state
                 .scoped_keys
                 .iter()
-                .filter(|(scope, _)| requested_scopes.contains(&(*scope).as_ref()))
+                .filter(|(scope, _)| valid_scopes.contains_key(*scope))
                 .map(|(key, value)| (key.clone(), value.clone()))
                 .collect();
-            let data = serde_json::to_string(&data)?;
+
+            // Return an error if we could not get the requested scopes from state
+            // Saves the client time instead of waiting for the server's response
+            let first_mismatch = scoped_keys
+                .iter()
+                .find(|(scope, _)| !valid_scopes.contains_key(*scope));
+            if let Some(scope) = first_mismatch {
+                return Err(ErrorKind::NoScopedKey(scope.0.clone()).into());
+            }
```

Comment:
I don't understand this check, or the comment about waiting for the server's response. Having an explicit testcase would probably help make it clearer.

It appears to be looking for a member of `scoped_keys` that is not in `valid_scopes`, but AFAICT every member of `scoped_keys` must be in `valid_scopes` by definition, because we check that as part of constructing `scoped_keys` in the first place.

You may find that doing this as a separate check is unnecessary if you invert the way you're building `scoped_keys` above: iterate over `valid_scopes`, call `self.state.get()` for each item, and allow it to throw the `NoScopedKey` error for you if we don't have that key in our local state.

Reply:
Yup that was a brain fart 🤦 Wanted to do the inverted check (i.e see if there is something in the `valid_scopes` not included) I like the inverted construction of `scoped_keys` better though! Thank you!